src/suggestion.c

```diff
@@ -0,0 +1,399 @@
+    if (new_prefix == NULL) {
+        return rc + EXIT_FAILURE;
+    }
+
+    strncpy(new_prefix, prefix, MAXLEN);
+
+    new_prefix[len] = suffix[0];
+
+    if (has_children(t, new_prefix) == EXIT_SUCCESS) {
+
+        // Move on to the next character, don't use up an edit
+        rc += suggestions(set, t, new_prefix, suffix + 1, edits_left, max_matches);
+    }
+
+    // Save some space now that we're done
+    free(new_prefix);
+
+    return rc + EXIT_SUCCESS;
+}
+
+// Helper function for suggestions that tries to remove the first character of the suffix
+int try_delete(match_t **set, trie_t *t, char *prefix, char *suffix, int edits_left, int max_matches) {
+
+    int rc = 0;
+
+    // Don't need to copy the string over as we aren't changing the prefix
+
+    if (has_children(t, prefix) == EXIT_SUCCESS) {
+
+        // Adding 1 to the suffix pointer will essentially delete the first character
+        rc += suggestions(set, t, prefix, suffix + 1, edits_left - 1, max_matches);
+    }
+
+    return rc + EXIT_SUCCESS;
+}
+
+// Helper function for suggestions that tries to replace the first character in the suffix and move it to the prefix
+int try_replace(match_t **set, trie_t *t, char *prefix, char *suffix, int edits_left, int max_matches) {
+
+    int i;
+    int rc = 0;
+    int len = strlen(prefix);
+    char *new_prefix;
+
+    // Ran out of space
+    if (len == MAXLEN - 1) {
+        return rc + EXIT_FAILURE;
+    }
+
+    for (i = 1; i < 248; i++) {
+
+        char c = (char)i;
+
+        if (trie_char_exists(t, c) == EXIT_SUCCESS) {
+
+            new_prefix = malloc(sizeof(char) * (MAXLEN + 1));
+            if (new_prefix == NULL) {
+                return rc + EXIT_FAILURE;
+            } else {
+
+                strncpy(new_prefix, prefix, MAXLEN);
+
+                // Try replacing the beginning of the suffix with each ASCII character
+                // And move that to the end of the prefix
+                new_prefix[len] = c;
+                new_prefix[len + 1] = '\0';
+
+                if (has_children(t, new_prefix) == EXIT_SUCCESS) {
+
+                    // Adding 1 to the suffix pointer will essentially delete the first character
+                    // Shifting the "replaced" character to the prefix
+                    rc += suggestions(set, t, new_prefix, suffix + 1, edits_left - 1, max_matches);
+                }
+
+                // Save some space now that we're done
+                free(new_prefix);
+            }
+        }
+    }
+
+    return rc + EXIT_SUCCESS;
+}
+
+// Helper function that tries to swap the last character of the prefix and first character of the prefix
+// and append both to the prefix
+int try_swap(match_t **set, trie_t * t, char *prefix, char *suffix, int edits_left, int max_matches) {
+
+    int rc = 0;
+    int len = strlen(prefix);
+    char* new_prefix;
+
+    if (len == 0 || strlen(suffix) == 0) {
+        return rc + EXIT_SUCCESS;
```

Comment:
Same as above.

Reply:
@dwahme

Reply:
Updated to return EXIT_SUCCESS and EXIT_FAILURE accordingly- this new structure should now make more sense